File: 957-minimum-add-to-make-parentheses-valid/minimum-add-to-make-parentheses-valid.java
class Solution {
    public int minAddToMakeValid(String s) {
        Stack<Character> st = new Stack<Character>();
        
        int res = 0;

        for(int i = 0;i<s.length();i++){
            if(s.charAt(i)=='('){
                st.push('(');
            }
            else{
                if(!st.empty()){
                    st.pop();
                }
                else{
                    res++;
                }
            }
        }

        if(!st.empty()){
            res += st.size();
        }

        return res;
    }
}

File: 957-minimum-add-to-make-parentheses-valid/minimum-add-to-make-parentheses-valid.cpp
class Solution {
public:
    int minAddToMakeValid(string s) {
        int open = 0;

        int res = 0;

        for(int i = 0;i<s.size();i++){
            if(s[i]=='('){
                open++;
            }
            else{
                if(open>0){
                    open--;
                }else{
                    res++;
                }
            }
        }

        if(open){
            res += open;
        }

        return res;
    }
};